### PR TITLE
Make tool parameter options capable of referring to metadata file

### PR DESCRIPTION
**Pull request by qiagu**
## What did you do? 
- Extend tool parameter options to generate dynamic options from metadata file(s) of a referred dataset or dataset collection. 
- Refactor linter for dynamic options.

## Why did you make this change?
- The `from_dataset` is able to generate dynamic options from the main file(s) of a dataset or dataset_collection, while `data_meta` in `filter` can refer to some types of metadata via attribute `key`, but file type is not one of them, as stated in #8599.

## How to test the changes? 
(select the most appropriate option; if the latter, provide steps for testing below)
- [x] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [ ] This is a refactoring of components with existing test coverage.
- [ ] Instructions for manual testing are as follows:
  1. [add testing steps and prerequisites here if you didn't write automated tests covering all your changes]

## License
- [x] I agree to license these contributions under [Galaxy's current license](https://github.com/galaxyproject/galaxy/blob/dev/LICENSE.txt).
- [x] I agree to allow the Galaxy committers to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT). If this condition is an issue, uncheck and just let us know why with an e-mail to galaxy-committers@lists.galaxyproject.org.

## For UI Components
- [ ] I've included a screenshot of the changes

### CODE REVIEW

**Comment by qiagu** [comment]
Not sure this is done properly. Any feedback is appreciated.

**Comment by jmchilton** [comment]
Can you include an example tool that demonstrates how to use this? Maybe something like was added in https://github.com/galaxyproject/galaxy/pull/2970/files.

**Comment by bernt-matthias** [comment]
Also documentation would be appreciated :)

**Comment by qiagu** [comment]
@jmchilton @bernt-matthias Added. Thanks!

**Comment by bernt-matthias** [comment]
Isnät this possible with https://docs.galaxyproject.org/en/master/dev/schema.html#tool-inputs-param-options-filter `data_meta` ?

**Comment by bernt-matthias** [comment]
I started working on extending this to metadata files, but this stayed unfinished: https://github.com/galaxyproject/galaxy/pull/8599

**Comment by jmchilton** [comment]
> Isn't this possible with https://docs.galaxyproject.org/en/master/dev/schema.html#tool-inputs-param-options-filter data_meta ?

This implementation seems nice and clean but I guess if there is already another way to do this we should just encourage that? @bernt-matthias are you saying this is possible now or would be possible with #8599? I wish I had a better handle on what is possible and what is best practice around dynamic options.

**Comment by bernt-matthias** [comment]
Let me just state that this is certainly a valuable contribution.

> are you saying this is possible now or would be possible with #8599? 

Its not possible with https://github.com/galaxyproject/galaxy/pull/8599 .. 

I reverted my initial work on metadata files at some point .. I summarized how far I got in this comment https://github.com/galaxyproject/galaxy/pull/8599#issuecomment-530877471

My main point is to simply use `key` instead of adding a new attribute. It should be possible to determine the type of the metadata automatically. I think there are string, list, mapping and files. At the moment only strings and lists are supported.

Further ideas/comments:

- if would be cool to support the cases where there referred dataset is a collection / multiple=true parameter.
- A potential problem is that one might want to limit how much of the metadata file is read .. or can we assume that these files are small (see https://github.com/galaxyproject/galaxy/pull/8599#issuecomment-693323484)

**Comment by qiagu** [comment]
Logically, the `data_meta` should support the file type metadata. That's what I looked for in the first place, but it didn't work out for me. I like the features naturally provided by `from_dataset`, such as, multiple `column`s (separate name and value), `startswith`, and potentially additional unique constraint using `filter`.  I see the `from_dataset` route was also mentioned in https://github.com/galaxyproject/galaxy/pull/8599. This implementation is very simple, maybe a little bit hacky, while I anticipate it may need a lot of effort to reorganize and extend `data_meta` filter to support all the aforementioned features. How about we keep `meta_file_key` for now? When `data_meta` is fully implemented, we then make `meta_file_key` depreciated. 
In my potential use case, the hyperperameters of machine learning models rarely go beyond 1MB in size.  If too-large metadata files do happen, I would consider simplifying the content of metadata from upstream. I understand the 1MB limit is necessary. Denial-of-service attack may happen otherwise. Besides, too many options may challenge the tool UI/UX, don't they?

>if would be cool to support the cases where there referred dataset is a collection / multiple=true parameter.
I think this is already supported, behaving the same as `from_dataset`?

File: lib/galaxy/tool_util/xsd/galaxy.xsd [hunk 1]
```diff
@@ -3519,6 +3533,11 @@ used to generate dynamic options.
         <xs:documentation xml:lang="en">Documentation for startswith</xs:documentation>
       </xs:annotation>
     </xs:attribute>
+    <xs:attribute name="meta_file_key" type="xs:string">
+      <xs:annotation>
+        <xs:documentation xml:lang="en">Documentation for meta_file_key</xs:documentation>
```
Comment (bernt-matthias) [inline lib/galaxy/tool_util/xsd/galaxy.xsd]: Can you replace this with actual docs?

**Comment by qiagu** [comment]
@jmchilton @bernt-matthias I wonder whether you have a clearer idea about this PR? Considering that the `filter` route may take long time to be implemented, I think the benefits of this extension outweigh the concerns. FYI, my another PR and many tools are waiting for this to be cleared. Thanks!

**Comment by bernt-matthias** [comment]
Hi @qiagu I'm not sure if we already have feature freeze for 21.01 .. I think @mvdbeek can tell us.  

Thinking again about it I think it might actually be a good idea to go your way. Also the syntax is just much nicer as I imagine it with the filter.

If I understand the code correctly I think it should also just work for the  case that the referred input is a collection of a data param with multiple=true. Could you add a test for this? I guess this is possible within the same test tool.

I also think that already only the first MB from the metadata file(s) is read. Could you check this .. ?

File: test/functional/tools/options_from_metadata_file.xml [hunk 2]
```diff
@@ -0,0 +1,29 @@
+<tool id="options_from_metadata_file" name="Test for options from dataset metadata file" version="1.0.0" profile="21.01">
+  <command><![CDATA[
+#assert $input.metadata.species_chromosomes
```
Comment (bernt-matthias) [inline test/functional/tools/options_from_metadata_file.xml]: Out of curiosity: What is the idea of this assertion?

**Comment by mvdbeek** [comment]
Freeze / branch is on the 3rd of May.

File: test/functional/tools/options_from_metadata_file.xml [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+echo '${species}' > '${output}'
+  ]]></command>
+  <inputs>
+    <param name="input" type="data" format="maf" label="MAF File"/>
+    <param name="species" type="select" optional="False" label="Select species for the input dataset" multiple="True">
```
Comment (bernt-matthias) [inline test/functional/tools/options_from_metadata_file.xml]:
```suggestion
    <param name="species" type="select" optional="false" label="Select species for the input dataset" multiple="true">
```

**Review by bernt-matthias — CHANGES_REQUESTED** [review]

File: lib/galaxy/tool_util/xsd/galaxy.xsd [hunk 4]
```diff
@@ -3430,6 +3430,20 @@ and demonstrates generating options from a dataset directly.
 </param>
 ```
 
+The ``meta_file_key`` can be used together with ``from_dataset``. In such cases,
```
Comment (bernt-matthias) [inline lib/galaxy/tool_util/xsd/galaxy.xsd]: Should we document that this works only with 21.01 / requires profile="21.01".

File: test/functional/tools/options_from_metadata_file.xml (hunk 2 above)
Comment (qiagu) [reply test/functional/tools/options_from_metadata_file.xml]: Copied from another test tool. It doesn't seem to be harmful, but I will be happy to remove it.

File: lib/galaxy/tool_util/xsd/galaxy.xsd (hunk 4 above)
Comment (qiagu) [reply lib/galaxy/tool_util/xsd/galaxy.xsd]: Sure. Will do.

**Comment by qiagu** [comment]
@bernt-matthias  
>I also think that already only the first MB from the metadata file(s) is read. Could you check this .. ?

I think so. The limit is also applied to the `from_dataset` alone.

File: lib/galaxy/tool_util/xsd/galaxy.xsd (hunk 1 above)
Comment (qiagu) [reply lib/galaxy/tool_util/xsd/galaxy.xsd]: Frankly, I'm not sure how to do this. I was a copycat. :)

Comment (bernt-matthias) [reply lib/galaxy/tool_util/xsd/galaxy.xsd]: Just put text there describing how the attribute works.

Also thinking if it's worth to extend the linter https://github.com/galaxyproject/galaxy/blob/b52c78ac1e7ebf36e4ca735e6c6a53838260f15c/lib/galaxy/tool_util/linters/inputs.py#L31

**Review by bernt-matthias — CHANGES_REQUESTED** [review]
Hi @qiagu I was so free to add a bit of linting to this PR .. would be great if someone could have a look. 

Missing is only the doc for the new attribute and the PR template.

**Comment by bernt-matthias** [comment]
Hmm. The test shows that only the first element of the collection is used.

File: lib/galaxy/tools/parameters/dynamic_options.py [hunk 5]
```diff
@@ -669,7 +670,9 @@ def get_fields(self, trans, other_values):
 
             options = []
             for dataset in datasets:
-                if not hasattr(dataset, 'file_name'):
+                if self.meta_file_key:
+                    dataset = getattr(dataset.metadata, self.meta_file_key, None)
```
Comment (bernt-matthias) [inline lib/galaxy/tools/parameters/dynamic_options.py]: I guess we should check this against `no_value` instead of `None`?

**Review by bernt-matthias — COMMENTED** [review]
A general comment on the example: The data used for the select that is taken from the `species_chromosomes` file is also contained in the `species` metadata element: 

https://github.com/galaxyproject/galaxy/blob/cf5ee5aed0cc164b6471a525e0a3da2b71ae4c8a/lib/galaxy/datatypes/sequence.py#L301

So there may be easier ways to implement this specific example.

File: lib/galaxy/tools/parameters/dynamic_options.py [hunk 6]
```diff
@@ -669,7 +670,9 @@ def get_fields(self, trans, other_values):
...
+                if not dataset or not hasattr(dataset, 'file_name'):
```
Comment (qiagu) [inline lib/galaxy/tools/parameters/dynamic_options.py]: I think this line can help check the `no_value`.

**Comment by bernt-matthias** [comment]
Seems that we have now some unrelated commits in this PR.

**Comment by qiagu** [comment]
Cleaned.

**Review by bernt-matthias — APPROVED** [review]
Since I co-authored a few commits it would be good if someone else could also review these changes.

In particular I'm unsure if the new lints should be errors / warning .. I have the feeling that the new ones are a bit more "aggressive" then usual (but maybe that is good).

File: lib/galaxy/tool_util/linters/inputs.py [hunk 7]
```diff
@@ -29,20 +30,75 @@ def lint_inputs(tool_xml, lint_ctx):
...
+            options = param.findall("./options")
+            filters = param.findall("./options/filter")
             select_options = param.findall('./option')
+
+            if dynamic_options is not None:
+                lint_ctx.warn(f"Select parameter [{param_name}] uses deprecated 'dynamic_options' attribute.")
+
+            # check if options are defined by exactly one possibility
+            if (dynamic_options is not None) + (len(options) > 0) + (len(select_options) > 0) != 1:
+                lint_ctx.error(f"Select parameter [{param_name}] options have to be defined by either 'option' tags, a 'option' tag or the attribute 'dynamic_options'.")
```
Comment (nsoranzo) [inline lib/galaxy/tool_util/linters/inputs.py]:
```suggestion
                lint_ctx.error(f"Select parameter [{param_name}] options have to be defined by either 'option' children elements, a 'options' element or the 'dynamic_options' attribute.")
```

File: lib/galaxy/tool_util/linters/inputs.py [hunk 8]
```diff
@@ -29,20 +30,72 @@ def lint_inputs(tool_xml, lint_ctx):
...
+                    lint_ctx.error(f"Select parameter [{param_name}] 'meta_file_key' is only compatible with 'from_dataset'.")
+
+                if options[0].get("options_filter_attribute", None) is not None:
+                    lint_ctx.warn(f"Select parameter [{param_name}] options uses deprecated 'options_filter_attribute' attribute.")
+
+                if options[0].get("transform_lines", None) is not None:
+                    lint_ctx.warn(f"Select parameter [{param_name}] options uses deprecated 'transform_lines' attribute.")
+
+            elif len(options) > 1:
+                lint_ctx.error(f"Select parameter [{param_name}] contains multiple options tags")
```
Comment (nsoranzo) [inline lib/galaxy/tool_util/linters/inputs.py]:
```suggestion
                lint_ctx.error(f"Select parameter [{param_name}] contains multiple options elements")
```

File: lib/galaxy/tool_util/xsd/galaxy.xsd [hunk 9]
```diff
@@ -3486,37 +3509,42 @@ used to generate dynamic options.
     </xs:sequence>
     <xs:attribute name="from_dataset" type="xs:string">
       <xs:annotation>
-        <xs:documentation xml:lang="en">Documentation for from_dataset</xs:documentation>
+        <xs:documentation xml:lang="en">Determine options from the data set given in the referred input parameter. If 'meta_file_key' is given the options are determined from the data in the metadata file of the input.</xs:documentation>
```
Comment (nsoranzo) [inline lib/galaxy/tool_util/xsd/galaxy.xsd]:
```suggestion
        <xs:documentation xml:lang="en">Determine options from the dataset given in the referred input parameter. If 'meta_file_key' is given, the options are determined from the data in the metadata file of the input.</xs:documentation>
```

**Comment by qiagu** [comment]
@nsoranzo Thanks for your comments!

**Review by mvdbeek — APPROVED** [review]

**Comment by qiagu** [comment]
@mvdbeek Thanks!

**Pull request merged**